Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebCore/platform/audio

cdumez opened the pull request:
#### 563f1bbd4b2ca0508561c9a562663a4fe0b12866
<pre>
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebCore/platform/audio
<a href="https://bugs.webkit.org/show_bug.cgi?id=285173">https://bugs.webkit.org/show_bug.cgi?id=285173</a>

Reviewed by Geoffrey Garen.

* Source/WebCore/Modules/webaudio/PannerNode.cpp:
(WebCore::PannerNode::processSampleAccurateValues):
* Source/WebCore/platform/audio/AudioArray.h:
(WebCore::AudioArray::~AudioArray):
(WebCore::AudioArray::resize):
(WebCore::AudioArray::span):
(WebCore::AudioArray::span const):
(WebCore::AudioArray::data):
(WebCore::AudioArray::data const):
(WebCore::AudioArray::size const):
(WebCore::AudioArray::isEmpty const):
(WebCore::AudioArray::at):
(WebCore::AudioArray::at const):
(WebCore::AudioArray::containsConstantValue const):
(): Deleted.
* Source/WebCore/platform/audio/AudioBus.cpp:
* Source/WebCore/platform/audio/AudioChannel.cpp:
* Source/WebCore/platform/audio/AudioChannel.h:
* Source/WebCore/platform/audio/AudioResampler.cpp:
(WebCore::AudioResampler::process):
* Source/WebCore/platform/audio/AudioResamplerKernel.cpp:
(WebCore::AudioResamplerKernel::process):
* Source/WebCore/platform/audio/AudioResamplerKernel.h:
* Source/WebCore/platform/audio/AudioSession.cpp:
(WebCore::convertEnumerationToString):
* Source/WebCore/platform/audio/AudioUtilities.cpp:
* Source/WebCore/platform/audio/Biquad.cpp:
(WebCore::Biquad::process):
(WebCore::Biquad::reset):
* Source/WebCore/platform/audio/DirectConvolver.cpp:
(WebCore::DirectConvolver::process):
* Source/WebCore/platform/audio/DownSampler.cpp:
(WebCore::DownSampler::process):
* Source/WebCore/platform/audio/DynamicsCompressor.cpp:
(WebCore::DynamicsCompressor::process):
(WebCore::DynamicsCompressor::setNumberOfChannels):
* Source/WebCore/platform/audio/DynamicsCompressor.h:
* Source/WebCore/platform/audio/DynamicsCompressorKernel.cpp:
(WebCore::DynamicsCompressorKernel::process):
* Source/WebCore/platform/audio/DynamicsCompressorKernel.h:
* Source/WebCore/platform/audio/EqualPowerPanner.cpp:
(WebCore::EqualPowerPanner::panWithSampleAccurateValues):
* Source/WebCore/platform/audio/EqualPowerPanner.h:
* Source/WebCore/platform/audio/FFTConvolver.cpp:
* Source/WebCore/platform/audio/HRTFElevation.cpp:
* Source/WebCore/platform/audio/HRTFKernel.cpp:
* Source/WebCore/platform/audio/HRTFPanner.cpp:
(WebCore::HRTFPanner::panWithSampleAccurateValues):
* Source/WebCore/platform/audio/HRTFPanner.h:
* Source/WebCore/platform/audio/IIRFilter.cpp:
(WebCore::evaluatePolynomial):
(WebCore::IIRFilter::process):
(WebCore::IIRFilter::getFrequencyResponse):
* Source/WebCore/platform/audio/Panner.h:
* Source/WebCore/platform/audio/PlatformMediaSession.cpp:
(WebCore::convertEnumerationToString):
* Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp:
* Source/WebCore/platform/audio/PlatformMediaSessionManager.h:
* Source/WebCore/platform/audio/PushPullFIFO.cpp:
* Source/WebCore/platform/audio/ReverbAccumulationBuffer.cpp:
(WebCore::ReverbAccumulationBuffer::accumulate):
* Source/WebCore/platform/audio/ReverbAccumulationBuffer.h:
* Source/WebCore/platform/audio/ReverbConvolverStage.cpp:
(WebCore::ReverbConvolverStage::process):
* Source/WebCore/platform/audio/ReverbInputBuffer.cpp:
* Source/WebCore/platform/audio/StereoPanner.cpp:
(WebCore::StereoPanner::panWithSampleAccurateValues):
* Source/WebCore/platform/audio/UpSampler.cpp:
(WebCore::UpSampler::process):
* Source/WebCore/platform/audio/ZeroPole.cpp:
(WebCore::ZeroPole::process):
* Source/WebCore/platform/audio/ZeroPole.h:

Canonical link: <a href="https://commits.webkit.org/288315@main">https://commits.webkit.org/288315@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0b8564e84493239b133a31e4802885c6d75830d2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82435 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2099 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36527 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87567 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33496 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84540 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2170 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9987 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64255 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22006 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85504 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1546 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75004 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44532 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1444 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29190 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32537 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/75423 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72736 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29825 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/88924 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/81489 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9741 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7004 "Found 2 new test failures: http/wpt/mediarecorder/record-96KHz-sources.html swipe/navigate-event-back-swipe-verify-ua-transition.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72654 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9967 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70818 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71871 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16010 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15016 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/1117 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12799 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9694 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15215 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/103902 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9568 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25202 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13034 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11338 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebCore/platform/audio/AudioArray.h:
```diff
@@ -67,38 +65,20 @@ class AudioArray {
-            CRASH();
-        m_size = n;
+        fastAlignedFree(m_allocation.data());
+        m_allocation = unsafeMakeSpan(static_cast<T*>(fastAlignedMalloc(alignment, initialSize)), n);
```
If we gave MallocSpan an alignedMalloc() function, we could drop the unsafeMakeSpan.

cdumez replied:
The tricky part is that `fastAlignedMalloc()` needs to be paired with `fastAlignedFree()` it seems.

cdumez replied:
I'll create a FastAlignedMalloc struct and use that instead of FastMalloc I think. Then we don't need to add an `alignedMalloc()` function to `MallocSpan` because `malloc()` would do the right thing.

cdumez replied:
Actually, we'll still need an explicit `alignedMalloc()` function because of the extra alignment parameter.

geoffreygaren replied:
FWIW, alignedFree() was only required by old versions of Windows, so we could potentially get rid of it at some point. But this is great as is.

cdumez committed:
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebCore/platform/audio
https://bugs.webkit.org/show_bug.cgi?id=285173

Reviewed by Geoffrey Garen.

* Source/WebCore/Modules/webaudio/PannerNode.cpp:
(WebCore::PannerNode::processSampleAccurateValues):
* Source/WebCore/platform/audio/AudioArray.h:
(WebCore::AudioArray::~AudioArray):
(WebCore::AudioArray::resize):
(WebCore::AudioArray::span):
(WebCore::AudioArray::span const):
(WebCore::AudioArray::data):
(WebCore::AudioArray::data const):
(WebCore::AudioArray::size const):
(WebCore::AudioArray::isEmpty const):
(WebCore::AudioArray::at):
(WebCore::AudioArray::at const):
(WebCore::AudioArray::containsConstantValue const):
(): Deleted.
* Source/WebCore/platform/audio/AudioBus.cpp:
* Source/WebCore/platform/audio/AudioChannel.cpp:
* Source/WebCore/platform/audio/AudioChannel.h:
* Source/WebCore/platform/audio/AudioResampler.cpp:
(WebCore::AudioResampler::process):
* Source/WebCore/platform/audio/AudioResamplerKernel.cpp:
(WebCore::AudioResamplerKernel::process):
* Source/WebCore/platform/audio/AudioResamplerKernel.h:
* Source/WebCore/platform/audio/AudioSession.cpp:
(WebCore::convertEnumerationToString):
* Source/WebCore/platform/audio/AudioUtilities.cpp:
* Source/WebCore/platform/audio/Biquad.cpp:
(WebCore::Biquad::process):
(WebCore::Biquad::reset):
* Source/WebCore/platform/audio/DirectConvolver.cpp:
(WebCore::DirectConvolver::process):
* Source/WebCore/platform/audio/DownSampler.cpp:
(WebCore::DownSampler::process):
* Source/WebCore/platform/audio/DynamicsCompressor.cpp:
(WebCore::DynamicsCompressor::process):
(WebCore::DynamicsCompressor::setNumberOfChannels):
* Source/WebCore/platform/audio/DynamicsCompressor.h:
* Source/WebCore/platform/audio/DynamicsCompressorKernel.cpp:
(WebCore::DynamicsCompressorKernel::process):
* Source/WebCore/platform/audio/DynamicsCompressorKernel.h:
* Source/WebCore/platform/audio/EqualPowerPanner.cpp:
(WebCore::EqualPowerPanner::panWithSampleAccurateValues):
* Source/WebCore/platform/audio/EqualPowerPanner.h:
* Source/WebCore/platform/audio/FFTConvolver.cpp:
* Source/WebCore/platform/audio/HRTFElevation.cpp:
* Source/WebCore/platform/audio/HRTFKernel.cpp:
* Source/WebCore/platform/audio/HRTFPanner.cpp:
(WebCore::HRTFPanner::panWithSampleAccurateValues):
* Source/WebCore/platform/audio/HRTFPanner.h:
* Source/WebCore/platform/audio/IIRFilter.cpp:
(WebCore::evaluatePolynomial):
(WebCore::IIRFilter::process):
(WebCore::IIRFilter::getFrequencyResponse):
* Source/WebCore/platform/audio/Panner.h:
* Source/WebCore/platform/audio/PlatformMediaSession.cpp:
(WebCore::convertEnumerationToString):
* Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp:
* Source/WebCore/platform/audio/PlatformMediaSessionManager.h:
* Source/WebCore/platform/audio/PushPullFIFO.cpp:
* Source/WebCore/platform/audio/ReverbAccumulationBuffer.cpp:
(WebCore::ReverbAccumulationBuffer::accumulate):
* Source/WebCore/platform/audio/ReverbAccumulationBuffer.h:
* Source/WebCore/platform/audio/ReverbConvolverStage.cpp:
(WebCore::ReverbConvolverStage::process):
* Source/WebCore/platform/audio/ReverbInputBuffer.cpp:
* Source/WebCore/platform/audio/StereoPanner.cpp:
(WebCore::StereoPanner::panWithSampleAccurateValues):
* Source/WebCore/platform/audio/UpSampler.cpp:
(WebCore::UpSampler::process):
* Source/WebCore/platform/audio/ZeroPole.cpp:
(WebCore::ZeroPole::process):
* Source/WebCore/platform/audio/ZeroPole.h:

Canonical link: https://commits.webkit.org/288315@main
